### PR TITLE
Match students with volunteers in waves

### PR DESCRIPTION
**Pull request by keerthanamuniraj**
Links
-----
Task: [https://www.notion.so/upchieve/Match-students-with-volunteers-in-waves-d061db2fc5a3476581532b2a8263a1c4](url)

Description
-----------
Twilio currently sends out message to all the volunteers who are available to help a student needing help.
This story does the following
Send out texts to volunteers in batches of 5 successively if no volunteer has joined the session.
The text is sent out to next batch after 30 seconds time delay.

Developer self-review checklist
-------------------------------
- [x] Task's requirements have been fully addressed
- [x] PR link has been posted in the task's comments
- [x] Potentially confusing code has been explained with comments
- [x] No warnings or errors have been introduced; all known error cases have been handled
- [x] Any appropriate documentation (within the code, README.md, docs, etc) has been updated
- [x] There are no new spelling/grammar mistakes in the UI, code, or documentation
- [ ] Branch has been deployed to staging, and all edge cases have been manually tested
- [x] Task and PR have been updated to show that this is ready for review

### CODE REVIEW

File: services/twilio.js [hunk 1]
```diff
@@ -56,32 +57,47 @@ var getAvailableVolunteersFromDb = function (subtopic) {
...
     getAvailableVolunteersFromDb(subtopic).exec(function (err, persons) {
-      persons.forEach(function (person) {
-        send(person.phone, person.firstname, subtopic)
+      var count = persons.count()
+      console.log(
+        `number of volunteers available` + count
+      )
+      //Send out invites in batches of 5 with 30 seconds delay
+      xrange(5).persons.forEach(function (person) {
+        var delayInMilliseconds = 30000
```
Comment (keerthanamuniraj) [inline services/twilio.js]: Added a decent time delay of 30 seconds. Please indicate if the time delay has to be increased/decreased.

File: package.json [hunk 2]
```diff
@@ -34,7 +34,8 @@
     "sendgrid": "^5.2.3",
     "socket.io": "^2.2.0",
     "twilio": "^3.28.1",
-    "validator": "^6.2.0"
+    "validator": "^6.2.0",
+    "xrange": "^1.0.1"
```
Comment (keerthanamuniraj) [inline package.json]: Need package xrange to create volunteer waves/batches

File: models/Session.js [hunk 3]
```diff
@@ -114,7 +117,13 @@ sessionSchema.methods.endSession = function (cb) {
   this.save(() => console.log(`Ended session ${this._id} at ${this.endedAt}`))
 }
 
-sessionSchema.methods.isActive = function (cb) {}
+sessionSchema.methods.isActive = function (cb) {
+  if(joinUser.volunteerJoinedAt){
```
Comment (keerthanamuniraj) [inline models/Session.js]: If the time at which a volunteer joined is known, then the session is considered active. Meaning, the text is subsequent volunteer batch should not be sent out.

File: services/twilio.js [hunk 4]
```diff
@@ -56,32 +57,47 @@ var getAvailableVolunteersFromDb = function (subtopic) {
 
   var query = User.find(userQuery)
     .select({ phone: 1, firstname: 1 })
-    .limit(5)
```
Comment (keerthanamuniraj) [inline services/twilio.js]: Looks like there was a hard limit set to 5 which was commented out to fetch the list of volunteer available matching the schedule.

**Comment by jkeat** [comment]
Implemented in other PRs

**Pull request closed without merging**